torch/quantization/ns/graph_passes.py

```diff
@@ -154,12 +154,19 @@ def _insert_dtype_cast_after_node(
     """
     dtype_cast_op = None
     dtype_cast_mod_cls = None
-    node_input_type_a = get_node_input_type(node_a, gm_a)
-    node_input_type_c = get_node_input_type(node_c, gm_b)
-
-    if node_input_type_a == NodeInputType.FP32 and node_input_type_c == NodeInputType.INT8:
+    node_input_type_a, _node_output_type_a = \
+        get_node_first_input_and_output_type(node_a, gm_a)
+    node_input_type_c, _node_output_type_c = \
+        get_node_first_input_and_output_type(node_c, gm_b)
+
+    if (
+        (node_input_type_a == NodeInputOrOutputType.FP32 and
+         node_input_type_c == NodeInputOrOutputType.INT8) or
+        (node_input_type_a == NodeInputOrOutputType.FP32 and
+         node_input_type_c == NodeInputOrOutputType.FP16)
```

Comment:
A general question, is this change for supporting Fakelowp ops? IIRC the output of Fakelowp ops is fp32 but casted to fp16 precision, so I guess it's for something else?

Reply:
yes, this is for fakelowp ops.  For example,

```
# original graph
x0 -> linear_fp32 -> x1

# graph with fakelowp replacement
x0_fp16 -> dequantize -> linear_fp32 -> fp32_to_fp16 -> x1_fp16

# graph with fp32 linear shadowing fakelowp linear
  dequantize -> linear_fp32 -> logger
    /
x0_fp16 -> dequantize -> linear_fp32 -> fp32_to_fp16 -> logger -> x1_fp16
```